aips/create_dip.py

```diff
@@ -0,0 +1,326 @@
+        aip_uuid=aip_uuid,
+        ss_url=ss_url,
+        ss_user_name=ss_user,
+        ss_api_key=ss_api_key,
+        directory=tmp_dir)
+
+    aip_file = am_client.download_aip()
+
+    if not aip_file:
+        LOGGER.error('Unable to download AIP')
+        return
+
+    LOGGER.info('Extracting AIP')
+    aip_dir = extract_aip(aip_file, aip_uuid, tmp_dir)
+
+    if not aip_dir:
+        return
+
+    LOGGER.info('Creating DIP')
+    dip_dir = create_dip(aip_dir, aip_uuid, output_dir)
+
+    if not dip_dir:
+        LOGGER.error('Unable to create DIP')
+        return
+
+    LOGGER.info('DIP created in: %s', dip_dir)
+
+
+def extract_aip(aip_file, aip_uuid, tmp_dir):
+    """Extract a downloaded AIP to a folder."""
+    command = ['7z', 'x', '-bd', '-y', '-o{0}'.format(tmp_dir), aip_file]
+    try:
+        subprocess.check_output(command, stderr=subprocess.STDOUT)
+    except subprocess.CalledProcessError as e:
+        LOGGER.error('Could not extract AIP, error: %s', e.output)
+        return
+
+    # Remove extracted file to avoid multiple entries with the same UUID
+    try:
+        os.remove(aip_file)
+    except OSError:
+        pass
+
+    # Find extracted entry. Assuming it contains the AIP UUID
+    for entry in os.listdir(tmp_dir):
+        if aip_uuid in entry:
+            extracted_entry = os.path.join(tmp_dir, entry)
+
+    if not extracted_entry:
+        LOGGER.error('Can not find extracted AIP by UUID')
+        return
+
+    # Return folder path if it's a directory
+    if os.path.isdir(extracted_entry):
+        return extracted_entry
+
+    # Re-try extraction if it's not a directory
+    return extract_aip(extracted_entry, aip_uuid, tmp_dir)
+
+
+def create_dip(aip_dir, aip_uuid, output_dir):
+    aip_name = os.path.basename(aip_dir)[:-37]
+    dip_dir = os.path.join(output_dir, '{}_{}_DIP'.format(aip_name, aip_uuid))
+    objects_dir = os.path.join(dip_dir, 'objects')
+    to_zip_dir = os.path.join(objects_dir, aip_name)
+
+    if os.path.exists(dip_dir):
+        LOGGER.warning('DIP folder already exists, overwriting')
+        shutil.rmtree(dip_dir)
+    os.makedirs(to_zip_dir)
+
+    LOGGER.info('Moving submissionDocumentation folder')
+    aip_sub_doc = '{}/data/objects/submissionDocumentation'.format(aip_dir)
+    if os.path.exists(aip_sub_doc):
+        to_zip_sub_doc = os.path.join(to_zip_dir, 'submissionDocumentation')
+        shutil.move(aip_sub_doc, to_zip_sub_doc)
+    else:
+        LOGGER.warning('submissionDocumentation folder not found')
+
+    LOGGER.info('Moving METS file')
+    aip_mets_file = '{}/data/METS.{}.xml'.format(aip_dir, aip_uuid)
+    if not os.path.exists(aip_mets_file):
+        LOGGER.error('Could not find AIP METS file')
+        return
+    to_zip_mets_file = '{}/METS.{}.xml'.format(to_zip_dir, aip_uuid)
+    shutil.move(aip_mets_file, to_zip_mets_file)
+
+    mets = metsrw.METSDocument.fromfile(to_zip_mets_file)
+    fsentries = mets.all_files()
+    for fsentry in fsentries:
```

Comment:
This stuff inside the loop is a good candidate to be refactored as maybe one or more functions. Not only easier to read but also easier to test, with clear dependencies for each function, etc...

Reply:
Right, thanks!